### DISCUSSION TITLE
cc65 just worked for the Apple 1

### DISCUSSION THREAD

**Discussion by davidly** [cc65/cc65#2123] [Show and tell]
In case anyone else is wondering about if/how cc65 works with an Apple 1...

Getting cc65 to work for the Apple 1 literally took a couple minutes. I build my app like this:

    cc65 -T -Oi -Or -t none ttt.c
    ca65 ttt.s
    ld65 -o ttt -t none ttt.o none.lib

Then create an Apple 1 compatible hex file using https://github.com/davidly/hd

    hd -w:0x1000 ttt >ttt.hex

Then run it in an Apple 1 emulator: https://github.com/davidly/ntvao

    ntvao -p -c ttt.hex

Then run it on the RetroTechLyfe Apple 1 Replica by sending the app via a serial port: https://github.com/davidly/ss

    ss /p:7 /s:ttt.hex /r:1000

To send data to the terminal on an Apple 1, do something like this:

    void show_int( val ) uint16_t val;
    {
        static uint8_t h, l;
        h = val >> 8;
        l = val;

        __asm__( "lda     %v", h );
        __asm__( "jsr     $ffdc" );
        __asm__( "lda     %v", l );
        __asm__( "jsr     $ffdc" );
    } /*show_int*/

    void show_char( val ) char val;
    {
        static uint8_t s_val;
        s_val = val;
        __asm__( "lda     %v", s_val );
        __asm__( "jsr     $ffef" );
    } /*show_char*/

    void show_string( str ) char * str;
    {
        while ( *str )
        {
            show_char( *str );
            str++;
        }
    } /*show_string*/

To exit cleanly back to the Apple 1 monitor, call this:

    void bye()
    {
        __asm__( "jsr     $ff1f" );
    } /*bye*/

I spent a day getting the Aztec 6502 compiler to work with the Apple 1, including modifying the C runtime
and adding startup assembly code. And the generated code for my app is > 3x slower than cc65.

**Comment by polluks** [comment]
You may add your knowledge to https://cc65.github.io/doc/intro.html#ss6.1.

**Comment by rofl0r** [comment]
would be nice if those tools you advertise were written in portable C, not MS-specific toy languages.

**Reply by davidly** [reply]
portable C is great. portable, open source C# is great too. hd referenced above works on Linux -- build and run C# using tools that have been open source for many years now. C# runs some of the largest services on the planet on both Linux and Windows. 

ss is Windows-specific. Feel free to contribute to make it cross-platform.
